src/tests/JIT/Regression/JitBlue/Runtime_77710/Runtime_77710.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+
+public class Runtime_77710
+{
+    public static int Main()
+    {
+        return Test(new Derived());
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static int Test(Base b)
+    {
+        b.M(callBase: false, recursing: false);
+        b.M(callBase: true, recursing: false);
+
+        int result = 100;
+
+        if (Base.BaseCalls != 1)
+        {
+            Console.WriteLine("Expected BaseCalls to be 1, is actually {0}", Base.BaseCalls);
+            result = -1;
+        }
+
+        if (Derived.DerivedCalls != 1)
+        {
+            Console.WriteLine("Expected DerivedCalls to be 1, is actually {0}", Derived.DerivedCalls);
+            result = -1;
+        }
+
+        return result;
+    }
+
+    private class Base
+    {
+        public static int BaseCalls;
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public virtual int M(bool callBase, bool recursing)
+        {
+            BaseCalls++;
+            return 0;
+        }
+    }
+
+    private class Derived : Base
+    {
+        public static int DerivedCalls;
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public override int M(bool callBase, bool recursing)
+        {
+            if (recursing)
+            {
+                DerivedCalls++;
+                return 0;
+            }
+
+            if (callBase)
+            {
+                base.M(false, true);
+            }
+            else
+            {
+                M(false, true);
+            }
+
+            return 0;
```

Comment:
@AndyAyersMS is it expected that tail merging does not kick in without this extra successor block? For
```csharp
        [MethodImpl(MethodImplOptions.NoInlining)]
        public override void M(bool callBase, bool recursing)
        {
            if (recursing)
            {
                DerivedCalls++;
                return;
            }

            if (callBase)
            {
                base.M(false, true);
            }
            else
            {
                M(false, true);
            }
         }
```

tail merging was not kicking in.

Reply:
I wonder if the tail call -> loop stuff runs even if the method is marked `NoInlining`?

Reply:
I don't think it's related to that, the IR looks like something we should be able to tail merge (assuming this bug isn't fixed). Here's the relevant jitdump, taken with a build before this fix. I would've expected to see tail merging.
```scala
*************** Finishing PHASE Compute preds
Trees after Compute preds

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..003)-> BB03 ( cond )                     i 
BB02 [0001]  1       BB01                  1       [003..010)        (return)                     i 
BB03 [0002]  1       BB01                  1       [010..013)-> BB05 ( cond )                     i 
BB04 [0003]  1       BB03                  1       [013..01C)        (return)                     i 
BB05 [0004]  1       BB03                  1       [01C..025)        (return)                     i 
-----------------------------------------------------------------------------------------------------------------------------------------

------------ BB01 [000..003) -> BB03 (cond), preds={} succs={BB02,BB03}

***** BB01
STMT00000 ( 0x000[E-] ... 0x001 )
               [000003] -----------                         ▌  JTRUE     void  
               [000002] -----------                         └──▌  EQ        int   
               [000000] -----------                            ├──▌  LCL_VAR   bool   V02 arg2         
               [000001] -----------                            └──▌  CNS_INT   int    0

------------ BB02 [003..010) (return), preds={BB01} succs={}

***** BB02
STMT00006 ( 0x003[E-] ... 0x00A )
               [000022] -A--G------                         ▌  ASG       int   
               [000021] ----G--N---                         ├──▌  FIELD     int    DerivedCalls
               [000020] ----G------                         └──▌  ADD       int   
               [000018] ----G------                            ├──▌  FIELD     int    DerivedCalls
               [000019] -----------                            └──▌  CNS_INT   int    1

***** BB02
STMT00007 ( 0x00F[E-] ... 0x00F )
               [000023] -----------                         ▌  RETURN    void  

------------ BB03 [010..013) -> BB05 (cond), preds={BB01} succs={BB04,BB05}

***** BB03
STMT00001 ( 0x010[E-] ... 0x011 )
               [000007] -----------                         ▌  JTRUE     void  
               [000006] -----------                         └──▌  EQ        int   
               [000004] -----------                            ├──▌  LCL_VAR   bool   V01 arg1         
               [000005] -----------                            └──▌  CNS_INT   int    0

------------ BB04 [013..01C) (return), preds={BB03} succs={}

***** BB04
STMT00004 ( 0x013[E-] ... 0x01B )
               [000016] --C-G------                         ▌  CALL      void   Base.M
               [000013] ----------- this                    ├──▌  LCL_VAR   ref    V00 this         
               [000014] ----------- arg1                    ├──▌  CNS_INT   int    0
               [000015] ----------- arg2                    └──▌  CNS_INT   int    1

***** BB04
STMT00005 ( 0x01B[E-] ... ??? )
               [000017] -----------                         ▌  RETURN    void  

------------ BB05 [01C..025) (return), preds={BB03} succs={}

***** BB05
STMT00002 ( 0x01C[E-] ... 0x024 )
               [000011] --C-G------                         ▌  CALLV vt-ind void   Base.M
               [000008] ----------- this                    ├──▌  LCL_VAR   ref    V00 this         
               [000009] ----------- arg1                    ├──▌  CNS_INT   int    0
               [000010] ----------- arg2                    └──▌  CNS_INT   int    1

***** BB05
STMT00003 ( 0x024[E-] ... ??? )
               [000012] -----------                         ▌  RETURN    void  

-------------------------------------------------------------------------------------------------------------------

*************** Starting PHASE Tail merge

*************** Finishing PHASE Tail merge [no changes]
```

Reply:
The algorithm looks for common predecessor sequences so it nees some sort of join point to get started.

There is no successor block for those returns, so no merging happens. I suppose we could treat all the `BBJ_RETURNs` as if they all had a common successor. Shouldn't be too hard to do...

Reply:
I see. Yeah, this might be a pretty common case where tail merging would kick in.